File: Data Structures & Algorithms/linked-list-cycle-detection/submission-0.js
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode} head
     * @return {boolean}
     */
    hasCycle(head) {
        let curr = head;
        const visitedNodes = new Set();
        while (curr !== null) {
            if (visitedNodes.has(curr)) {
                return true;
            }
            visitedNodes.add(curr);
            curr = curr.next;
        }
        return false;
    }
}

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        length = 0
        curr = head
        while curr is not None:
            length += 1
            curr = curr.next
        
        index = length - n
        if index == 0:
            return head.next
            
        prev = head
        for i in range(index - 1):
            prev = prev.next
        
        if prev.next is not None:
            prev.next = prev.next.next
        return head
        